### PR TITLE
Speed up dashboard loading

### PR DESCRIPTION
**Pull request by we-ai**
This PR update data fetching and UI renderings, to solve issue [#23](https://github.com/episphere/dashboard/issues/23). Below are details:
- Combined 14 separated calls into 1 call, to get metric data from backend.
- For speed, saved metric data are used for a period of time (currently 20 minutes) before doing a new data fetch. There's no need to fetch new data for each dashboard refresh, since backend metric data are updated every 1 hour (in prod).
- Extracted repeated code handling data transformation into a function `transformDataForCharts`, to simplify code for dashboard re-rendering.
- There're 3 different functions having the same name `dropdownTrigger`. Used different names to distinguish them.
- Other updates improving code readability or speed.

### CODE REVIEW

File: siteManagerDashboard/index.js [hunk 1]
```diff
@@ -197,42 +206,11 @@ const renderActivityCheck = () => {
...
-    return divElement;
- }
-
 const renderDashboard = async () => {
     if (localStorage.dashboard || await getIdToken()) {
         animation(true);
-        const siteKey = await getAccessToken();
-        const isAuthorized = await authorize(siteKey);
+        const accessToken = await getAccessToken();
+        const isAuthorized = await authorize(accessToken);
```
Comment (JoeArmani) [inline siteManagerDashboard/index.js:213]: Nice naming differentiation! We use site___ for many things, which can get confusing.

**Review by JoeArmani — APPROVED** [review]
looks good, just a few minor questions/comments.

File: siteManagerDashboard/index.js [hunk 2]
```diff
@@ -61,11 +62,20 @@ window.onload = async () => {
     activityCheckController();
     const userSession = localStorage.getItem('userSession');
     userSession && appState.setState({ userSession: JSON.parse(userSession) });
-}
+
+    const statsData = localStorage.getItem("statsData");
+    const statsDataUpdateTime = localStorage.getItem("statsDataUpdateTime");
```
[the remark this answers is not in the document]
Comment (we-ai) [reply siteManagerDashboard/index.js:67]: These 2 variables are updated together to have matched values. 
For a new user session, `statsData` (line 66) is null,  `statsDataUpdateTime` (line 67 will) is also null (without needing to check). Then condition check in line 68 (`if (statsData && Date.now() - parseInt(statsDataUpdateTime) < statsDataTimeLimit)`) will be false, thus line 71 is executed.
Yes, the fallback is line 71.

[the remark this answers is not in the document]
Comment (JoeArmani) [reply siteManagerDashboard/index.js:67]: makes sense to me, thanks for the explanation

File: siteManagerDashboard/index.js [hunk 3]
```diff
@@ -37,6 +37,7 @@ const datadogConfig = {
 }
 
 const isLocalDev = location.hostname === 'localhost' || location.hostname === '127.0.0.1';
+const statsDataTimeLimit = 1200000; // 20 minutes
```
Comment (abhinavjonnada82) [inline siteManagerDashboard/index.js:40]: Like this appproach.

**Comment by abhinavjonnada82** [comment]
Looks good & can move forward

**Review by abhinavjonnada82 — APPROVED** [review]

File: siteManagerDashboard/index.js [hunk 4]
```diff
@@ -256,128 +234,152 @@ const renderDashboard = async () => {
...
+    } else {
+      const response = await getStatsForDashboard(accessToken);
+      if (response.code !== 200) return;
+      
+      [statsData] = response.data;
+      if (!statsData || Object.keys(statsData).length === 0) return;
 
-        animation(false);
+      appState.setState({statsData, statsDataUpdateTime: currentTime});
+      localStorage.setItem("statsData", JSON.stringify(statsData));
```
Comment (anthonypetersen) [inline siteManagerDashboard/index.js:252]: just curious, why use app state and local storage?

**Review by anthonypetersen — APPROVED** [review]
looks good, couple of questions asked but nothing major

File: siteManagerDashboard/index.js (hunk 4 above)
Comment (we-ai) [reply siteManagerDashboard/index.js:252]: appState is quick, but cannot carry data across page loads. Local storage can save data across page refreshed and loads, but is slower. Better combine the strength of them.

Comment (anthonypetersen) [reply siteManagerDashboard/index.js:252]: good to know!

File: siteManagerDashboard/index.js [hunk 5]
```diff
@@ -256,128 +234,152 @@ const renderDashboard = async () => {
...
-            }
-        })
-
-    }
-}
-
+};
+
+const handleSiteSelection = (siteTextContent) => {
+  const siteSelectionButton = document.getElementById("dropdownSites");
```
[the remark this answers is not in the document]
Comment (we-ai) [reply siteManagerDashboard/index.js:375]: Not really.  `handleSiteSelection`  is called after `renderSiteKeyList()` which has `id="dropdownSites" ` as part of its returned HTML string.

File: siteManagerDashboard/participantChartsRender.js [hunk 6]
```diff
@@ -1,265 +1,143 @@
...
+        <div class = "col-lg-4 charts">
+            <div class="col-lg-12 sub-div-shadow viz-div" id="ssnMetrics"></div>
+        </div>
+        <div class = "col-lg-4 charts">
+            <div class="col-lg-12 sub-div-shadow viz-div" id="biospecimenMetrics"></div>
+        </div>
+    </div>
+    `;
+
+    const mainContent = document.getElementById('mainContent');
```
[the remark this answers is not in the document]
Comment (we-ai) [reply siteManagerDashboard/participantChartsRender.js:72]: No really. It is defined in entry file index.html line 48 (`<div id="mainContent"></div>`).

**Pull request merged**